### PR TITLE
Board

### PR DESCRIPTION
**Pull request by ImMPrada**
This PR upgrades Chess::Core::Board

A Board will have these responsibilities:
- build the Cells
- occupy the cells with pieces
- know which pieces are on the board
- enable the ability to check if any piece can move to an indicated cell
- enable the ability to know if there is a checkmate

### CODE REVIEW

File: lib/chess/core/board.rb [hunk 1]
```diff
@@ -1,127 +1,119 @@
...
-          (MIN_INDEX..MAX_INDEX).each do |row_index|
-            @cells[column.to_sym]
-
-            cell = @cells[column.to_sym][row_index]
-            next unless cell.occuped?
-
-            occuped_cells[cell.team] << cell.coordinates.to_a
+        def all
+          to_a.reject do |piece|
+            piece.nil? || piece.captured? || piece.symbol == :K
```
Comment (mapra99) [inline lib/chess/core/board.rb:29]:
```suggestion
        def available_pieces
          to_a.reject do |piece|
            piece.nil? || piece.captured? || piece.symbol == :K
```

File: lib/chess/core/board.rb [hunk 2]
```diff
@@ -1,127 +1,119 @@
...
+        ]
       end
 
-      def find_cell(cell_algebraic_string)
-        splitted_cell_string = cell_algebraic_string.split('')
-        @cells[splitted_cell_string[0].to_sym][splitted_cell_string[1].to_i - 1]
+      def create_pawns(team, pawns_index)
+        (0..7).to_a.map do |column_index|
+          Core::Pieces::Pawn.create_and_occupy(team, @cells[pawns_index][column_index])
+        end
```
Comment (mapra99) [inline lib/chess/core/board.rb]: I think there is no need to convert the range to array
```suggestion
        (0..7).map do |column_index|
          Core::Pieces::Pawn.create_and_occupy(team, @cells[pawns_index][column_index])
        end
```

File: lib/chess/functional/target_cell_moves.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+          all_pieces[king_in_check.enemies_team].all
+        )
+        return false if can_any_friend_intersect_path?(
+          check_pice, king_in_check,
+          all_pieces[king_in_check.team].all,
+          cells
+        )
+
+        true
+      end
```
Comment (mapra99) [inline lib/chess/functional/target_cell_moves.rb]: typo
```suggestion
      def checkmate?(check_piece, king_in_check, all_pieces, cells)
        return false if can_any_piece_move_to?(
          check_pice.current_cell,
          cells,
          all_pieces[king_in_check.team].all
        )
        return false if king_can_escape?(
          king_in_check,
          cells,
          all_pieces[king_in_check.enemies_team].all
        )
        return false if can_any_friend_intersect_path?(
          check_pice, king_in_check,
          all_pieces[king_in_check.team].all,
          cells
        )

        true
      end
```

File: spec/lib/chess/core/board_spec.rb [hunk 4]
```diff
@@ -0,0 +1,167 @@
+require 'spec_helper'
+require 'chess/core/board'
+require 'chess/core/cell'
+
+RSpec.describe Chess::Core::Board do
+  describe '#create_and_occupy' do
```
Comment (mapra99) [inline spec/lib/chess/core/board_spec.rb]:
```suggestion
  describe '.create_and_occupy' do
```

File: spec/lib/chess/core/board_spec.rb [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      it 'returns false' do
+        expect(board.checkmate?(check_pice, king_in_check, board.pieces, board.cells)).to be(false)
+      end
+    end
+
+    describe 'when a friend piece, can intersect the enemy path' do
+      before { hardcode_pieces_checkmate_case4(board) }
+
+      it 'returns false' do
+        expect(board.checkmate?(check_pice, king_in_check, board.pieces, board.cells)).to be(false)
```
Comment (mapra99) [inline spec/lib/chess/core/board_spec.rb]: typo
```suggestion
        expect(board.checkmate?(check_piece, king_in_check, board.pieces, board.cells)).to be(false)
```

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -1,5 +1,6 @@
 require 'simplecov'
 require 'simplecov-lcov'
+require_relative 'helper/board_pieces'
```
Comment (mapra99) [inline spec/spec_helper.rb]: don't expand unnecessarily the dependencies here. If this helper is being used only on a few spec files, import it on each file, not here

**Review by mapra99 — APPROVED** [review]
go go go

File: lib/chess/core/board.rb (hunk 2 above)
Comment (ImMPrada) [reply lib/chess/core/board.rb]: that's right

File: lib/chess/functional/target_cell_moves.rb (hunk 3 above)
Comment (ImMPrada) [reply lib/chess/functional/target_cell_moves.rb]: fixed

File: spec/lib/chess/core/board_spec.rb (hunk 5 above)
Comment (ImMPrada) [reply spec/lib/chess/core/board_spec.rb]: done

File: spec/lib/chess/core/board_spec.rb (hunk 4 above)
Comment (ImMPrada) [reply spec/lib/chess/core/board_spec.rb]: done

File: spec/spec_helper.rb (hunk 6 above)
Comment (ImMPrada) [reply spec/spec_helper.rb]: done

**Pull request merged**